### PR TITLE
Update adapter.py conform preCICE v1.4.1 and Code_Aster v14.4

### PR DESCRIPTION
**Pull request by micheltakken**
This version of the adapter supports preCICE v1.4.1 and Code_Aster v14.4 on an Ubuntu 18.04 system

### CODE REVIEW

File: cht/adapter.py [hunk 1]
```diff
@@ -2,307 +2,296 @@
...
+		self.interfaces = []
+		self.numInterfaces = len(config)
+		self.MESH = MESH
+		self.MODEL = MODEL
+		self.MAT = MAT
+		self.LOADS = []
+		self.isNonLinear = isNonLinear
+		self.participantName = participantName
+		self.preciceDt = -1
+		self.precice = precice.Interface(participantName, 0, 1)
```
Comment (MakisH) [inline cht/adapter.py]: Is there a reason for changing the indentation everywhere? If not, please keep the previous one.

File: cht/adapter.py [hunk 2]
```diff
@@ -2,307 +2,296 @@
...
-        self.preciceDt = -1
-        self.precice = PySolverInterface(participantName, 0, 1)
-        self.precice.configure(preciceConfigFile)
-        self.configure(config)
-    
-    def configure(self, config):
-        L = [None] * self.numInterfaces        # Loads
-        SM = [None] * self.numInterfaces    # Shifted meshes
-        for i in range(self.numInterfaces):
-            # Shifted mesh (interface nodes displaced by a distance delta in the direction of the surface normal
```
Comment (MakisH) [inline cht/adapter.py]: Why removing this comment?

File: cht/adapter.py [hunk 3]
```diff
@@ -2,307 +2,296 @@
...
+
+		# In Code_Aster, write-data is located at the nodes
+		self.nodesMeshName = names["write-mesh"]
+		# and read-data is located at the face centers
+		self.faceCentersMeshName = names["read-mesh"]
+
+		self.computeNormals()
+
+		self.nodeCoordinates = np.array([p for p in self.SHMESH.sdj.COORDO.VALE.get()])
+		# Casted len(...)/3 to int
```
Comment (MakisH) [inline cht/adapter.py]: You can remove this comment now.

File: cht/adapter.py [hunk 4]
```diff
@@ -2,307 +2,296 @@
...
+			},
+		)
+		N = CREA_CHAMP(
+			MODELE=DUMMY,
+			TYPE_CHAM='NOEU_GEOM_R',
+			GROUP_MA=self.groupName,
+			OPERATION='NORMALE'
+		)
+		self.normals = N.EXTR_COMP().valeurs
+		# Casted len(...)/3 to int
```
Comment (MakisH) [inline cht/adapter.py]: You can also remove this comment.

File: cht/adapter.py (hunk 3 above)
Comment (MakisH) [reply cht/adapter.py]: After discussing, I see the problem with rounding. Please fix this before merging.

File: cht/adapter.py (hunk 1 above)
Comment (micheltakken) [reply cht/adapter.py]: Python needs a correct indentation to run correctly. Somehow I got an error for incorrect indentation for the previous one, therefore I have redone the indentation of the entire code.

File: cht/adapter.py (hunk 2 above)
Comment (micheltakken) [reply cht/adapter.py]: These comments seemed unnecessary to me. If you think they're helpful, I will reinsert them.

File: cht/adapter.py (hunk 3 above)
Comment (micheltakken) [reply cht/adapter.py]: In the original code (python 2.x) the division is done with an "integer division", meaning that the result of x/3 will be floored. In python3 the division of two integers is a true division, returning a double. To get the desired outcome, an additional casting to `int` is necessary after the division operation.
The comments will be removed.

File: cht/adapter.py (hunk 4 above)
Comment (micheltakken) [reply cht/adapter.py]: Comment removed.

File: cht/adapter.py (hunk 1 above)
Comment (MakisH) [reply cht/adapter.py]: It does need tabs, but somehow the size of tabs (in spaces) was changed. This is something you can configure in your editor (e.g. in Atom: `Edit > Preferences > Editor > Tab Length`).

Comment (MakisH) [reply cht/adapter.py]: Please ignore my above comments and keep your indentation.

File: README.md [hunk 5]
```diff
@@ -1,5 +1,9 @@
...
 
-This adapter was implemented as part of the [master thesis of Lucia Cheung](https://www5.in.tum.de/pub/Cheung2016_Thesis.pdf) in cooperation with [SimScale](https://www.simscale.com/). Currently only conjugate heat transfer is supported. 
+This adapter was implemented as part of the [master thesis of Lucia Cheung](https://www5.in.tum.de/pub/Cheung2016_Thesis.pdf) in cooperation with [SimScale](https://www.simscale.com/). Currently only conjugate heat transfer is supported.
 
-A documentation is available in the `cht` subfolder. 
+The adapter currently supports preCICE v1.4.1 and Code_Aster v14.4. The following tutorials are available.
+
+* Flow over plate OpenFoam-Code_Aster ([wiki](https://github.com/precice/code_aster-adapter/wiki/Flow-over-plate-Code_Aster-Tutorial))
+
+A documentation is available in the `cht` subfolder, and in the [wiki](wiki).
```
Comment (MakisH) [inline README.md]:
```suggestion
A documentation is available in the `cht` subfolder, and in the [wiki](https://github.com/precice/code_aster-adapter/wiki).
```

File: cht/adapter.comm [hunk 6]
```diff
@@ -134,11 +134,11 @@ while adapter.isCouplingOngoing():
...
         if k % settings["simulation"]["output-frequency"] == 0:
-               filename = "output-" + str(k) + ".rmed"
-               DEFI_FICHIER(FICHIER=filename, UNITE=80)
-               IMPR_RESU(FORMAT='MED', RESU=_F(RESULTAT=TEMP))
-               call(["mv", filename, settings["base-path"] + "/" + participant["directory"]])
-               DEFI_FICHIER(ACTION='LIBERER', UNITE=80)
+               filename = "REPE_OUT/output-" + str(k) + ".rmed"
+               DEFI_FICHIER(FICHIER=filename, UNITE=80+k, TYPE='BINARY')
+               IMPR_RESU(FORMAT='MED', UNITE=80+k, RESU=_F(RESULTAT=TEMP))
+               #call(["mv", filename, settings["base-path"] + "/" + participant["directory"]])
```
Comment (MakisH) [inline cht/adapter.comm]: Why is this commented out? What does it do? Why not remove it completely?

File: tutorials/buoyantSimpleFoam-aster/Allclean.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+rm -f precice-Fluid-convergence.log
+rm -f precice-Fluid-events.json
+rm -f precice-Fluid-iterations.log
+rm -f solid/solid.mess
+rm -f solid/solid.resu
+
+# Add file deletion for solid.rmed
+
+# pyFoamClearCase.py
```
Comment (MakisH) [inline tutorials/buoyantSimpleFoam-aster/Allclean.sh]: We don't need this line:

```suggestion
```

We can instead use the intrinsic OpenFOAM `cleanCase` script, if we first load it using:

```bash
# Source tutorial clean functions
. $WM_PROJECT_DIR/bin/tools/CleanFunctions
```

See an [example](https://github.com/precice/openfoam-adapter/blob/master/tutorials/CHT/flow-over-plate/buoyantPimpleFoam-laplacianFoam/Allclean).

File: tutorials/buoyantSimpleFoam-aster/Allclean.sh [hunk 8]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+rm -f precice-Fluid-convergence.log
+rm -f precice-Fluid-events.json
+rm -f precice-Fluid-iterations.log
+rm -f solid/solid.mess
+rm -f solid/solid.resu
+
+# Add file deletion for solid.rmed
```
Comment (MakisH) [inline tutorials/buoyantSimpleFoam-aster/Allclean.sh]: Todo

File: tutorials/buoyantSimpleFoam-aster/config.comm [hunk 9]
```diff
@@ -0,0 +1,34 @@
+settings = \
+{'base-path': '/home/michel/Git/code_aster-adapter/tutorials/buoyantSimpleFoam-aster',
```
Comment (MakisH) [inline tutorials/buoyantSimpleFoam-aster/config.comm]:
```suggestion
{'base-path': 'path/to/code_aster-adapter/tutorials/buoyantSimpleFoam-aster',
```

File: tutorials/buoyantSimpleFoam-aster/Allclean.sh (hunk 7 above)
Comment (MakisH) [reply tutorials/buoyantSimpleFoam-aster/Allclean.sh]: There is actually already an `Allclean` file inside the `Fluid/` case. In the rest of the tutorials we have the `Allclean` script in the same path as the `Allrun`.

File: tutorials/buoyantSimpleFoam-aster/Allclean.sh [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
```
Comment (MakisH) [inline tutorials/buoyantSimpleFoam-aster/Allclean.sh]: Rename this file from `Allclean.sh` to `Allclean` for consistency.

File: tutorials/buoyantSimpleFoam-aster/fluid/Allclean [hunk 11]
```diff
@@ -0,0 +1,2 @@
+. $WM_PROJECT_DIR/bin/tools/CleanFunctions
+cleanCase
```
Comment (MakisH) [inline tutorials/buoyantSimpleFoam-aster/fluid/Allclean]: Remove this file and move the functionality into the `Allclean` in the parent directory.

File: tutorials/buoyantSimpleFoam-aster/config.comm (hunk 9 above)
Comment (micheltakken) [reply tutorials/buoyantSimpleFoam-aster/config.comm]: Set this environment in the Allrun script.

Comment (MakisH) [reply tutorials/buoyantSimpleFoam-aster/config.comm]: For example: `export CA_BASE_PATH=$(pwd)`

File: cht/adapter.comm (hunk 6 above)
Comment (micheltakken) [reply cht/adapter.comm]: Removed it completely

File: tutorials/buoyantSimpleFoam-aster/config.comm (hunk 9 above)
Comment (micheltakken) [reply tutorials/buoyantSimpleFoam-aster/config.comm]: For now pushed a code with the 'path/to/' solution

File: tutorials/buoyantSimpleFoam-aster/Allclean.sh (hunk 7 above)
Comment (MakisH) [reply tutorials/buoyantSimpleFoam-aster/Allclean.sh]: The file was renamed to `Allclean` and everything looks good there.

File: tutorials/buoyantSimpleFoam-aster/config.comm (hunk 9 above)
Comment (MakisH) [reply tutorials/buoyantSimpleFoam-aster/config.comm]: Maybe you can try accessing the environment variable using the `os` Python package: `os.environ['CA_BASE_PATH']`. See [StackOverflow](https://stackoverflow.com/questions/4906977/how-to-access-environment-variable-values).

Comment (micheltakken) [reply tutorials/buoyantSimpleFoam-aster/config.comm]: I accessed the environment in adapter.comm with `os.environ['CA_BASE_PATH']`

File: cht/adapter.py [hunk 12]
```diff
@@ -2,307 +2,292 @@
...
+		self.writeDataSize = 0
+
+		self.MESH = MESH
+		# Shifted mesh (contains only the interface, and is shifted by delta in the direction opposite to the normal)
+		self.SHMESH = SHMESH
+		self.MODEL = MODEL
+		self.MAT = MAT
+		self.LOAD = None
+		self.LOADS = []
+		self.mesh = partition.MAIL_PY()
```
Comment (MakisH) [inline cht/adapter.py]: Why was this needed?

Comment (micheltakken) [reply cht/adapter.py]: Because the imports were changed (see line 7). Utilai is now imported as a whole. To get the contents of MAIL_PY(), we need to call it by partition.MAIL_PY()
